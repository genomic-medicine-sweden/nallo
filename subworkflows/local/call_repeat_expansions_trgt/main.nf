include { TRGT_GENOTYPE    } from '../../../modules/nf-core/trgt/genotype/main'
include { SAMTOOLS_INDEX   } from '../../../modules/nf-core/samtools/index/main'
include { SAMTOOLS_SORT    } from '../../../modules/nf-core/samtools/sort/main'
include { SAMTOOLS_CONVERT } from '../../../modules/nf-core/samtools/convert/main'
include { ADD_FOUND_IN_TAG } from '../../../modules/local/add_found_in_tag/main'
include { BCFTOOLS_SORT    } from '../../../modules/nf-core/bcftools/sort/main'
include { TRGT_MERGE       } from '../../../modules/nf-core/trgt/merge/main'
include { BCFTOOLS_INDEX   } from '../../../modules/nf-core/bcftools/index/main'

workflow CALL_REPEAT_EXPANSIONS_TRGT {
    take:
    ch_bam_bai  // channel: [mandatory] [ val(meta), path(bam), path(bai) ]
    ch_fasta    // channel: [mandatory] [ val(meta), path(fasta) ]
    ch_fai      // channel: [mandatory] [ val(meta), path(fai) ]
    ch_bed      // channel: [mandatory] [ val(meta), path(bed) ]
    cram_output // bool: Publish alignments as CRAM (true) or BAM (false)

    main:

    ch_versions = Channel.empty()

    ch_bam_bai
        .map { meta, bam, bai -> [meta, bam, bai, meta.sex == 1 ? 'XY' : 'XX'] }
        .set { ch_trgt_input }

    // Run TRGT
    TRGT_GENOTYPE(
        ch_trgt_input,
        ch_fasta,
        ch_fai,
        ch_bed,
    )
    ch_versions = ch_versions.mix(TRGT_GENOTYPE.out.versions)

    // Sort and index bam
    SAMTOOLS_SORT(
        TRGT_GENOTYPE.out.bam,
        [[], []],
    )
    ch_versions = ch_versions.mix(SAMTOOLS_SORT.out.versions)

    SAMTOOLS_INDEX(SAMTOOLS_SORT.out.bam)
    ch_versions = ch_versions.mix(SAMTOOLS_INDEX.out.versions)

    // Publish spanning reads as CRAM if requested
    if (cram_output) {
        SAMTOOLS_CONVERT(
            SAMTOOLS_SORT.out.bam.join(SAMTOOLS_INDEX.out.bai, failOnDuplicate: true, failOnMismatch: true),
            ch_fasta,
            ch_fai,
        )
        ch_versions = ch_versions.mix(SAMTOOLS_CONVERT.out.versions)
    }

    // Add FOUND_IN=TRGT tag
    ADD_FOUND_IN_TAG(
        TRGT_GENOTYPE.out.vcf.map { meta, vcf -> [meta, vcf, []] },
        "TRGT",
    )

    // Sort and index bcf
    BCFTOOLS_SORT(ADD_FOUND_IN_TAG.out.vcf)
    ch_versions = ch_versions.mix(BCFTOOLS_SORT.out.versions)

    BCFTOOLS_SORT.out.vcf
        .join(BCFTOOLS_SORT.out.tbi, failOnMismatch: true, failOnDuplicate: true)
        .map { meta, bcf, csi -> [[id: meta.family_id], bcf, csi] }
        .groupTuple()
        .set { ch_trgt_merge_in }

    TRGT_MERGE(
        ch_trgt_merge_in,
        [[], []],
        [[], []],
    )
    ch_versions = ch_versions.mix(TRGT_MERGE.out.versions)

    BCFTOOLS_INDEX(
        TRGT_MERGE.out.vcf
    )
    ch_versions = ch_versions.mix(BCFTOOLS_INDEX.out.versions)

    emit:
    sample_vcf = BCFTOOLS_SORT.out.vcf // channel: [ val(meta), path(vcf) ]
    sample_tbi = BCFTOOLS_SORT.out.tbi // channel: [ val(meta), path(tbi) ]
    family_vcf = TRGT_MERGE.out.vcf // channel: [ val(meta), path(vcf) ]
    family_tbi = BCFTOOLS_INDEX.out.tbi // channel: [ val(meta), path(tbi) ]
    sample_bam = SAMTOOLS_SORT.out.bam // channel: [ val(meta), path(bam) ]
    sample_bai = SAMTOOLS_INDEX.out.bai // channel: [ val(meta), path(bai) ]
    versions   = ch_versions // channel: [ versions.yml ]
}
