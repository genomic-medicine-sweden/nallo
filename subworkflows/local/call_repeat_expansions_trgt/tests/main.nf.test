// TODO nf-core: Once you have added the required tests, please run the following command to build this file:
// nf-core subworkflows test call_repeat_expansions_trgt
nextflow_workflow {

    name "Test Subworkflow CALL_REPEAT_EXPANSIONS_TRGT"
    script "../main.nf"
    workflow "CALL_REPEAT_EXPANSIONS_TRGT"

    tag "subworkflows"
    tag "subworkflows_"
    tag "subworkflows/call_repeat_expansions_trgt"
    // TODO nf-core: Add tags for all modules used within this subworkflow. Example:
    tag "samtools"
    tag "samtools/sort"
    tag "samtools/index"


    // TODO nf-core: Change the test name preferably indicating the test-data and file-format used
    test("sarscov2 - bam - single_end") {

        when {
            workflow {
                """
                // TODO nf-core: define inputs of the workflow here. Example:
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                ]
                input[1] = [
                    [ id:'genome' ],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
                //TODO nf-core: Add all required assertions to verify the test output.
            )
        }
    }
}
