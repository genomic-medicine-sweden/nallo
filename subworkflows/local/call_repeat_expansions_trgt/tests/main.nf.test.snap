{
    "[bam, bai], fasta, fai, bed": {
        "content": [
            "7e217a420f6cf91acb089797e7e8dafe",
            "880f8d47f964d3c8541d6acbca5ba02e",
            "fda458478c19ccf3b6b5fc128cecb534",
            [
                "versions.yml:md5,1b2027c02a323f942fc47f46cd82bd2e",
                "versions.yml:md5,2ce6af1042b39b1a4bdf881e43144c95",
                "versions.yml:md5,38462d5f29b6b0e354a5e305017ed80d",
                "versions.yml:md5,7158c6b7f001f4ca9aba515e8082e698",
                "versions.yml:md5,ad3480a461a9352bfe3344244cc01ab0",
                "versions.yml:md5,f388b35aae2ba0122406d46d8e3ca76c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-06-23T11:40:36.706942091"
    },
    "[bam, bai], fasta, fai, bed - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "family_id": "family",
                            "sex": 1
                        },
                        "test_sorted.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "family_id": "family",
                            "sex": 1
                        },
                        "test_sorted.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "family"
                        },
                        "family.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "family"
                        },
                        "family.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "family_id": "family",
                            "sex": 1
                        },
                        "test.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "family_id": "family",
                            "sex": 1
                        },
                        "test.bam.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    "versions.yml:md5,1b2027c02a323f942fc47f46cd82bd2e",
                    "versions.yml:md5,2ce6af1042b39b1a4bdf881e43144c95",
                    "versions.yml:md5,38462d5f29b6b0e354a5e305017ed80d",
                    "versions.yml:md5,7158c6b7f001f4ca9aba515e8082e698",
                    "versions.yml:md5,ad3480a461a9352bfe3344244cc01ab0",
                    "versions.yml:md5,f388b35aae2ba0122406d46d8e3ca76c"
                ],
                "family_tbi": [
                    [
                        {
                            "id": "family"
                        },
                        "family.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "family_vcf": [
                    [
                        {
                            "id": "family"
                        },
                        "family.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "sample_bai": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "family_id": "family",
                            "sex": 1
                        },
                        "test.bam.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "sample_bam": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "family_id": "family",
                            "sex": 1
                        },
                        "test.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "sample_tbi": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "family_id": "family",
                            "sex": 1
                        },
                        "test_sorted.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "sample_vcf": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "family_id": "family",
                            "sex": 1
                        },
                        "test_sorted.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1b2027c02a323f942fc47f46cd82bd2e",
                    "versions.yml:md5,2ce6af1042b39b1a4bdf881e43144c95",
                    "versions.yml:md5,38462d5f29b6b0e354a5e305017ed80d",
                    "versions.yml:md5,7158c6b7f001f4ca9aba515e8082e698",
                    "versions.yml:md5,ad3480a461a9352bfe3344244cc01ab0",
                    "versions.yml:md5,f388b35aae2ba0122406d46d8e3ca76c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-06-23T11:40:47.541212623"
    }
}
